src/translate.py

```diff
@@ -1062,27 +1091,34 @@ def format(self, fmt: Formatter) -> str:
             return f"MIPS2C_FIELD({var.format(fmt)}, {Type.ptr(self.type).format(fmt)}, {offset_str})"
         else:
             prefix = "unk" + ("_" if fmt.coding_style.unknown_underscore else "")
-            field_name = prefix + format_hex(self.offset)
+            field_path = [prefix + format_hex(self.offset)]
 
-        if isinstance(var, AddressOf):
-            if isinstance(var.expr, GlobalSymbol) and var.expr.array_dim is not None:
-                needs_deref = True
-            else:
-                needs_deref = False
-                var = var.expr
-        else:
-            needs_deref = True
+        # Rewrite `x[0].y` to `x->y` by dropping the initial `0` in field_path
+        if (
+            field_path
+            and len(field_path) >= 2
+            and field_path[0] == 0
+            and isinstance(field_path[1], str)
+        ):
+            field_path = field_path[1:]
```

Comment:
While I've convinced myself that this is correct, it feels sketchy to strip a part of the field path like this. The reason it's correct is that access_path_to_field_name does the semantically wrong thing when the path starts with a number: the natural thing to do given e.g. [1, "foo"] would be to output `x->[1].foo` (which is invalid C but the natural reading is `(*x)[1].foo`), but instead it strips the `->` and outputs `x[1].foo`. This stripping is done precisely when it's not done here, so the same amount of stripping happens. Not sure off hand how to best structure this, a comment may be helpful if nothing else.